versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -0,0 +1,935 @@
+    Map<Key, ContentsId> keyToContentsIds = new HashMap<>();
+    Set<ContentsId> contentsIds = new HashSet<>();
+    try (Stream<CommitLogEntry> log =
+        takeUntilExcludeLast(commitLogFetcher(ctx, refHead), e -> remainingKeys.isEmpty())) {
+      log.peek(entry -> entry.getDeletes().forEach(remainingKeys::remove))
+          .flatMap(entry -> entry.getPuts().stream())
+          .filter(put -> remainingKeys.remove(put.getKey()))
+          .filter(put -> keyFilter.check(put.getKey(), put.getContentsId(), put.getType()))
+          .forEach(
+              put -> {
+                nonGlobal.put(put.getKey(), put.getValue());
+                keyToContentsIds.put(put.getKey(), put.getContentsId());
+                contentsIds.add(put.getContentsId());
+              });
+    }
+
+    Map<ContentsId, ByteString> globals = fetchGlobalStates(ctx, contentsIds);
+
+    return nonGlobal.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                Entry::getKey,
+                e ->
+                    ContentsAndState.of(
+                        e.getValue(), globals.get(keyToContentsIds.get(e.getKey())))));
+  }
+
+  /**
+   * Fetches the global-state information for the given content-ids.
+   *
+   * @param ctx technical context
+   * @param contentIds the contents-ids to fetch
+   * @return map of content-id to state
+   */
+  protected abstract Map<ContentsId, ByteString> fetchGlobalStates(
+      OP_CONTEXT ctx, Set<ContentsId> contentIds);
+
+  /**
+   * Write a new commit-entry, the given commit entry is to be persisted as is. All values of the
+   * given {@link CommitLogEntry} can be considered valid and consistent.
+   *
+   * <p>Implementations however can enforce strict consistency checks/guarantees, like a best-effort
+   * approach to prevent hash-collisions but without any other consistency checks/guarantees.
+   */
+  protected abstract void writeIndividualCommit(OP_CONTEXT ctx, CommitLogEntry entry)
+      throws ReferenceConflictException;
+
+  /**
+   * Write multiple new commit-entries, the given commit entries are to be persisted as is. All
+   * values of the * given {@link CommitLogEntry} can be considered valid and consistent.
+   *
+   * <p>Implementations however can enforce strict consistency checks/guarantees, like a best-effort
+   * approach to prevent hash-collisions but without any other consistency checks/guarantees.
+   */
+  protected abstract void writeMultipleCommits(OP_CONTEXT ctx, List<CommitLogEntry> entries)
+      throws ReferenceConflictException;
+
+  /**
+   * Check whether the commits in the range {@code sinceCommitExcluding] .. [upToCommitIncluding}
+   * contain any of the given {@link Key}s.
+   *
+   * <p>Conflicts are reported via {@code mismatches}.
+   */
+  protected void checkConflictingKeysForCommit(
+      OP_CONTEXT ctx,
+      Hash upToCommitIncluding,
+      Hash sinceCommitExcluding,
+      Set<Key> keys,
+      boolean[] sinceSeen,
```

Comment:
why pass a mutable `boolean[1]` instead of returning a `boolean`

Reply:
my guess when reviewing the code was so that the value can be modified

Reply:
it appears tp be just a return value in the current usage. its modified in a lambda scope in the method but i don't see why the boolean array couldn't be declared in the method and just returning the element at the end of the method couldn't work. Maybe i am jus tnot reading it correclty?

Reply:
That `sinceSeen` value is computed within a Stream-lambda, so it can't be simply returned. Updated the code a little.